src/compiler/scala/tools/nsc/typechecker/Infer.scala

```diff
@@ -447,27 +448,35 @@ trait Infer extends Checkable {
      *  @return map from tparams to inferred arg, if inference was successful, tparams that map to None are considered left undetermined
      *    type parameters that are inferred as `scala.Nothing` and that are not covariant in `restpe` are taken to be undetermined
      */
-    def adjustTypeArgs(tparams: List[Symbol], tvars: List[TypeVar], targs: List[Type], restpe: Type = WildcardType): AdjustedTypeArgs.Result  = {
-      val buf = AdjustedTypeArgs.Result.newBuilder[Symbol, Option[Type]]
+    def adjustTypeArgs(tparams: List[Symbol], tvars: List[TypeVar], targs: List[Type], restpe: Type = WildcardType): AdjustedTypeArgs  = {
+      val okParams = ListBuffer[Symbol]()
+      val okArgs = ListBuffer[Type]()
+      val undetParams = ListBuffer[Symbol]()
+      val allArgs = ListBuffer[Type]()
 
       foreach3(tparams, tvars, targs) { (tparam, tvar, targ) =>
         val retract = (
               targ.typeSymbol == NothingClass                                         // only retract Nothings
           && (restpe.isWildcard || !varianceInType(restpe)(tparam).isPositive)  // don't retract covariant occurrences
         )
 
-        buf += ((tparam,
-          if (retract) None
-          else Some(
-            if (targ.typeSymbol == RepeatedParamClass)     targ.baseType(SeqClass)
+        if (retract) {
+          undetParams += tparam
+          allArgs += NothingTpe
+        } else {
+          val arg =
+            if (targ.typeSymbol == RepeatedParamClass) targ.baseType(SeqClass)
             else if (targ.typeSymbol == JavaRepeatedParamClass) targ.baseType(ArrayClass)
             // this infers Foo.type instead of "object Foo" (see also widenIfNecessary)
             else if (targ.typeSymbol.isModuleClass || tvar.constr.avoidWiden) targ
             else targ.widen
-          )
-        ))
+          okParams += tparam
+          okArgs += arg
+          allArgs += arg
+        }
       }
-      buf.result()
+
+      new AdjustedTypeArgs(tparams, okParams.toList, okArgs.toList, undetParams.toList, allArgs.toList)
```

Comment:
```suggestion
      new AdjustedTypeArgs(okParams.toList, okArgs.toList, undetParams.toList, allArgs.toList)
```
It's a case class, and we do not need to return input `tparams`.